deepchecks/tabular/checks/model_evaluation/weak_segments_performance.py

```diff
@@ -0,0 +1,328 @@
+
+        if context.features_importance is not None:
+            feature_rank = context.features_importance.sort_values(ascending=False).keys()
+            feature_rank = np.asarray([col for col in feature_rank if col in encoded_dataset.features])
+        else:
+            feature_rank = np.asarray(encoded_dataset.features)
+
+        scorer = context.get_single_scorer(self.user_scorer)
+        weak_segments = self._weak_segments_search(dummy_model, encoded_dataset, feature_rank,
+                                                   loss_per_sample, scorer)
+        if len(weak_segments) == 0:
+            raise DeepchecksProcessError('Unable to train a error model to find weak segments.')
+        avg_score = round(scorer(dummy_model, encoded_dataset), 3)
+
+        display = self._create_heatmap_display(dummy_model, encoded_dataset, weak_segments, avg_score,
+                                               scorer) if context.with_display else []
+
+        for idx, segment in weak_segments.copy().iterrows():
+            if segment['Feature1'] in encoded_dataset.cat_features:
+                weak_segments['Feature1 range'][idx] = \
+                    self._format_partition_vec_for_display(segment['Feature1 range'], segment['Feature1'], None)[0]
+            if segment['Feature2'] in encoded_dataset.cat_features:
+                weak_segments['Feature2 range'][idx] = \
+                    self._format_partition_vec_for_display(segment['Feature2 range'], segment['Feature2'], None)[0]
+
+        display_msg = 'Showcasing intersections of features with weakest detected segments.<br>Can ' \
+                      'observe the full list of weak segments in the check result value. '
+        return CheckResult({'weak_segments_list': weak_segments, 'avg_score': avg_score, 'scorer_name': scorer.name},
+                           display=[display_msg, DisplayMap(display)])
+
+    def _target_encode_categorical_features_fill_na(self, dataset: Dataset) -> Dataset:
+        values_mapping = defaultdict(list)
+        df_aggregated = dataset.features_columns.copy()
+        for col in dataset.cat_features:
+            categories_to_mask = [k for k, v in df_aggregated[col].value_counts().items() if
+                                  v / dataset.n_samples < self.categorical_aggregation_threshold]
+            df_aggregated.loc[np.isin(df_aggregated[col], categories_to_mask), col] = 'other'
+
+        if len(dataset.cat_features) > 0:
+            t_encoder = TargetEncoder(cols=dataset.cat_features)
+            df_encoded = t_encoder.fit_transform(df_aggregated, dataset.label_col)
+            for col in dataset.cat_features:
+                values_mapping[col] = pd.concat([df_encoded[col], df_aggregated[col]], axis=1).drop_duplicates()
+        else:
+            df_encoded = df_aggregated
+        self.encoder_mapping = values_mapping
+        df_encoded.fillna(df_encoded.mode(axis=0), inplace=True)
```

Comment:
Why fillna with mode for numericals as well? The common practice is to use mean for numericals.
Also, for categoricals, you can also just use "None" as a category, and if it's not common it would go to "other" anyway.
Anyway, I don't think it's right to do this **after** the target encoding - as you distort the information gained by the target encoding this way (no longer necessarily represents the correct value)